core/block_context.go

```diff
@@ -0,0 +1,87 @@
+package core
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/contract_comm/blockchain_parameters"
+	"github.com/celo-org/celo-blockchain/contract_comm/currency"
+	gpm "github.com/celo-org/celo-blockchain/contract_comm/gasprice_minimum"
+	"github.com/celo-org/celo-blockchain/core/types"
+	"github.com/celo-org/celo-blockchain/core/vm"
+)
+
+// BlockContext represents contextual information about the blockchain state
+// for a given block
+type BlockContext interface {
+	// GetGoldGasPriceMinimum retrieves the gas price minimum for the CELO token
+	GetGoldGasPriceMinimum() *big.Int
+
+	// GetGasPriceMinimum retrieves the gas price minimum for any currency
+	// Also indicates if the currency is not whitelisted
+	GetGasPriceMinimum(feeCurrency *common.Address) (gpm *big.Int, isWhitelisted bool)
+
+	// GetIntrinsicGasForAlternativeFeeCurrency retrieves intrisic gas to be paid for
+	// any tx with a non native fee currency
+	GetIntrinsicGasForAlternativeFeeCurrency() uint64
+}
+
+// defaultBlockContext is the default implementation of BlockContext
+type defaultBlockContext struct {
+	goldGasPriceMinimum     *big.Int
+	nonGoldGasPriceMinimums map[common.Address]*big.Int
+
+	gasForAlternativeCurrency uint64
+}
+
+// NewBlockContext creates a block context for a given block (represented by the
+// header & state).
+// state MUST be pointing to header's stateRoot
+func NewBlockContext(header *types.Header, state vm.StateDB) BlockContext {
+	gasForAlternativeCurrency := blockchain_parameters.GetIntrinsicGasForAlternativeFeeCurrency(header, state)
+
+	whitelistedCurrencies, err := currency.CurrencyWhitelist(header, state)
+	if err != nil {
+		whitelistedCurrencies = []common.Address{}
+	}
+
+	goldGasPriceMinimum, err := gpm.GetGasPriceMinimum(nil, header, state)
+	_ = err // Ignore the error since gpm.GetGasPriceMinimum returns the Fallback value on error
+
+	nonGoldGasPriceMinimums := make(map[common.Address]*big.Int, len(whitelistedCurrencies))
+	for _, currency := range whitelistedCurrencies {
+		gpm, err := gpm.GetGasPriceMinimum(&currency, header, state)
```

Comment:
It looks like this returns an error in fewer places than we might want and will result in accepting alternate currencies even if, for example, the gas price minimum contract is not deployed https://github.com/celo-org/celo-blockchain/blob/0f7f773a5dc9d34ba032f61d405011a95aa3856a/contract_comm/gasprice_minimum/gasprice_minimum.go#L125-L130